Add phase table model

jhaigh0 opened the pull request:
### Description of work

Part of the MVP workshop week.

Adds `PhaseTableModel` which holds an FDA context and some other state, previously in the presenter.
Some methods have been moved to the Model if they were only interacting with the context.
Tests updated, and a new test file created for tests only concerning methods now in the model.

Fixes #38104

### To test:

Code review and https://developer.mantidproject.org/Testing/MuonAnalysis_test_guides/Muon_FDA.html#id5

*This does not require release notes* because **refactoring change**

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jhaigh0 committed:
create phase table model

jhaigh0 committed:
move model only tests to a new file

rbauststfc commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/phase_table_widget/phase_table_model.py:
```diff
@@ -0,0 +1,132 @@
+
+    @property
+    def phasequad_calculation_thread(self):
+        return self._calculation_thread
```
Should these `phasequad_calculation_thread` properties be using the `self._phasequad_calculation_thread` variable?

rbauststfc commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/phase_table_widget/phase_table_presenter.py:
```diff
@@ -89,13 +77,13 @@ def update_current_run_list(self):
         """Cancels the current algorithm if executing"""
-        if self.current_alg is not None:
-            self.current_alg.cancel()
+        if self.model.current_alg is not None:
```
Just a thought, but looking at where `self.model.current_alg` is used, instead of exposing properties for `current_alg` could the model potentially just provide a `cancel_current_alg()` method that does all of this, plus a `clear_current_alg()` method that simply sets `self._current_alg = None`. Maybe that way we encapsulate just a bit more of the management of `current_alg` in the model?

rbauststfc commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/phase_table_widget/phase_table_presenter.py:
```diff
@@ -217,60 +181,60 @@ def handle_calculate_phase_table_clicked(self):
                 if phasequad.name == name:
                     self.add_phasequad_to_analysis(False, False, phasequad)
-                    self.context.group_pair_context.remove_phasequad(phasequad)
+                    self.model.group_pair_context.remove_phasequad(phasequad)
```
Is it worth wrapping this in a `remove_phasequad` method on the model, as it seems to be used in two places?

rbauststfc commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/phase_table_widget/phase_table_presenter.py:
```diff
@@ -217,60 +181,60 @@ def handle_calculate_phase_table_clicked(self):
         # Calculate the new table in a separate thread
-        self.calculation_thread = self.create_phase_table_calculation_thread()
-        self.calculation_thread.threadWrapperSetUp(
+        self.model.calculation_thread = self.create_phase_table_calculation_thread()
```
I'm wondering if data calculations like this one and the phasequad calculation on line 275 might be better in the model, assuming they work OK there?

jhaigh0 replied:
I think I left it there so it would be able to easily add the `self.handle_phase_table_calculation*` methods.

rbauststfc replied:
Yes, I guess I was thinking that technically you might be able to pass the callback methods through to the model in the same way that they're currently being passed to a method on the calculation thread. Maybe this is one to leave for Rob's advice, though, if we ask him to gatekeeper this.

jhaigh0 replied:
@robertapplin what do you think?

robertapplin replied:
I think Rachel's idea sounds good, to pass the callback methods through to the model, assuming its simple enough to do this. But if its a lot of work, then its not so bad to keep it the same

rbauststfc replied:
Thanks @robertapplin, I've had a go at doing that. I think you could probably open an issue to look at adding a bit more test coverage to the new model, but hopefully this will do for now.

@jhaigh0 I'm happy with all the changes you've made on this. Once all the tests have run through, perhaps you could review just the bits I've refactored in the last commit? If that all looks OK to you then I'll approve this and it can be gatekeepered. I've run through the manual tests since my refactor to check it's working OK, and all looks fine to me.

rbauststfc reviewed (commented):
Thanks for this, looks great and is functionally all working well when I ran the manual tests through.

I noticed one typo in the phasequad thread properties, and had a couple of other thoughts when looking through. Let me know what you think. If you feel like it might be worth trying any of them out then I'm very happy to pick up doing that, depending on what you prefer.

jhaigh0 committed:
pr suggested changes

rbauststfc committed:
Move performing calculations on a thread to model

Move the code that runs the phasequad and phase table calculations on a
new thread into the model

jhaigh0 commented:
Happy with the changes made in the latest commit that move the threading operations to the model 👍

rbauststfc reviewed (approved):
I've followed the manual testing steps since all changes have been made and everything's working well. All PR comments have been addressed, either by myself or Jonathan, and we've both approved each others changes, so this is ready to be gatekeepered.

robertapplin commented:
Thanks for your hard work with this PR! The code separation between the Presenter and Model is much improved, and the Model also has its own test suite now.
